File: iOS/Summit/Summit/Http/Requester/AttendRequester.swift
//
//  AttendRequester.swift
//  Summit
//
//  Created by Leapfrog-Software on 2018/04/29.
//  Copyright © 2018年 Leapfrog-Inc. All rights reserved.
//

import Foundation

struct AttendData {
    
    let tableId: String
    var userIds: [String]
    
    init?(data: Dictionary<String, Any>) {
        
        guard let tableId = data["tableId"] as? String else {
            return nil
        }
        self.tableId = tableId
        self.userIds = data["userIds"] as? Array<String> ?? []
    }
}

struct ChatData {
    
    let id: String
    let senderId: String
    let tableId: String
    let datetime: String
    let chat: String
    
    init?(data: Dictionary<String, Any>) {
        
        guard let id = data["id"] as? String else {
            return nil
        }
        self.id = id
        self.senderId = data["senderId"] as? String ?? ""
        self.tableId = data["tableId"] as? String ?? ""
        self.datetime = data["datetime"] as? String ?? ""
        self.chat = data["chat"] as? String ?? ""
    }
}

class AttendRequester {
    
    var scheduleId = ""
    var chatList = [ChatData]()
    var attendList = [AttendData]()
    
    func initialize(scheduleId: String) {
        self.scheduleId = scheduleId
    }
    
    func attend(tableId: String, completion: @escaping ((Bool) -> ())) {
        
        let params = [
            "command": "attend",
            "scheduleId": self.scheduleId,
            "userId": SaveData.shared.userId,
            "tableId": tableId
        ]
        ApiManager.post(params: params) { [weak self] result, data in
            if result, let datas = data as? Dictionary<String, Any> {
                if let chats = datas["chats"] as? Array<Dictionary<String, Any>>,
                    let attends = datas["attends"] as? Array<Dictionary<String, Any>> {
                    self?.chatList = chats.compactMap { ChatData(data: $0) }
                    self?.attendList = attends.compactMap { AttendData(data: $0) }
                    completion(true)
                    return
                }
            }
            completion(false)
        }
    }
    
    func postChat(chatId: String, tableId: String, chat: String, completion: @escaping ((Bool) -> ())) {
        
        let params = [
            "command": "postChat",
            "scheduleId": self.scheduleId,
            "chatId": chatId,
            "senderId": SaveData.shared.userId,
            "tableId": tableId,
            "chat": chat
        ]
        ApiManager.post(params: params) { result, _ in
            completion(result)
        }
    }
}

File: Android/app/src/main/java/leapfrog_inc/summit/Http/Requester/AttendRequester.java
package leapfrog_inc.summit.Http.Requester;

import org.json.JSONArray;
import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

import leapfrog_inc.summit.Function.Constants;
import leapfrog_inc.summit.Function.SaveData;
import leapfrog_inc.summit.Http.HttpManager;

/**
 * Created by Leapfrog-Software on 2018/05/08.
 */

public class AttendRequester {

    public static class AttendData {

        public String tableId;
        public ArrayList<String> userIds = new ArrayList<String>();

        static public AttendData create(JSONObject json) {

            try {
                String tableId = json.getString("tableId");
                JSONArray userIds = json.getJSONArray("userIds");

                AttendData attendData = new AttendData();
                attendData.tableId = tableId;
                for (int i = 0; i < userIds.length(); i++) {
                    attendData.userIds.add(userIds.getString(i));
                }
                return attendData;

            } catch(Exception e) {}

            return null;
        }
    }

    public static class ChatData {

        public String id;
        public String senderId;
        public String tableId;
        public Calendar datetime;
        public String chat;

        static public ChatData create(JSONObject json) {

            try {
                String id = json.getString("id");
                String senderId = json.getString("senderId");
                String tableId = json.getString("tableId");

                String datetimeStr = json.getString("datetime");
                SimpleDateFormat format = new SimpleDateFormat("yyyyMMddkkmmss");
                format.setTimeZone(TimeZone.getTimeZone("Asia/Tokyo"));
                Date datetime = format.parse(datetimeStr);
                Calendar calendar = Calendar.getInstance();
                calendar.setTime(datetime);

                String chat = json.getString("chat");

                ChatData chatData = new ChatData();
                chatData.id = id;
                chatData.senderId = senderId;
                chatData.tableId = tableId;
                chatData.datetime = calendar;
                chatData.chat = chat;
                return chatData;

            } catch(Exception e) {}

            return null;
        }
    }

    private String mScheduleId;
    private ArrayList<AttendData> mAttendList = new ArrayList<AttendData>();
    private ArrayList<ChatData> mChatList = new ArrayList<ChatData>();

    public void initialize(String scheduleId) {
        mScheduleId = scheduleId;
    }

    public void attend(String tableId, final AttendRequesterCallback callback){

        mAttendList.clear();
        mChatList.clear();

        HttpManager httpManager = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {
                if (result) {
                    try {
                        JSONObject jsonObject = new JSONObject(data);
                        String ret = jsonObject.getString("result");
                        if (ret.equals("0")) {
                            JSONArray attends = jsonObject.getJSONArray("attends");
                            for (int i = 0; i < attends.length(); i++) {
                                AttendData attendData = AttendData.create(attends.getJSONObject(i));
                                if (attendData != null) {
                                    mAttendList.add(attendData);
                                }
                            }
                            JSONArray chats = jsonObject.getJSONArray("chats");
                            for (int i = 0; i < chats.length(); i++) {
                                ChatData chatData = ChatData.create(chats.getJSONObject(i));
                                if (chatData != null) {
                                    mChatList.add(chatData);
                                }
                            }
                            callback.didReceiveData(true);
                            return;
                        }
                    } catch(Exception e) {}
                }
                callback.didReceiveData(false);
            }
        });
        StringBuffer param = new StringBuffer();
        param.append("command=attend");
        param.append("&");
        param.append(("scheduleId=" + mScheduleId));
        param.append("&");
        param.append(("tableId=" + tableId));
        param.append("&");
        param.append(("userId=" + SaveData.getInstance().userId));
        httpManager.execute(Constants.ServerApiUrl, "POST", param.toString());
    }

    public interface AttendRequesterCallback {
        void didReceiveData(boolean result);
    }

    public void postChat(String chatId, String tableId, String chat, final PostChatCallback callback) {

        HttpManager httpManager = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {
                if (result) {
                    try {
                        JSONObject jsonObject = new JSONObject(data);
                        String ret = jsonObject.getString("result");
                        if (ret.equals("0")) {
                            callback.didReceiveData(true);
                            return;
                        }
                    } catch(Exception e) {}
                }
                callback.didReceiveData(false);
            }
        });
        StringBuffer param = new StringBuffer();
        param.append("command=postChat");
        param.append("&");
        param.append(("scheduleId=" + mScheduleId));
        param.append("&");
        param.append(("chatId=" + chatId));
        param.append("&");
        param.append(("tableId=" + tableId));
        param.append("&");
        param.append(("senderId=" + SaveData.getInstance().userId));
        param.append("&");
        param.append(("chat=" + chat));
        httpManager.execute(Constants.ServerApiUrl, "POST", param.toString());
    }

    public interface PostChatCallback {
        void didReceiveData(boolean result);
    }

    public ArrayList<ChatData> queryChatList(int tableIndex) {

        ArrayList<ChatData> ret = new ArrayList<ChatData>();
        for (int i = 0; i < mChatList.size(); i++) {
            ChatData chatData = mChatList.get(i);
            if (chatData.tableId.equals(String.format("%d", tableIndex))) {
                ret.add(chatData);
            }
        }
        return ret;
    }

    public ArrayList<AttendData> getAttendList() {
        return mAttendList;
    }

    public ArrayList<String> queryAttendUserIds(int tableIndex) {

        for (int i = 0; i < mAttendList.size(); i++) {
            AttendData attendData = mAttendList.get(i);
            if (attendData.tableId.equals(String.format("%d", tableIndex))) {
                return attendData.userIds;
            }
        }
        return new ArrayList<String>();
    }
}

